plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/model/CodeWhispererModel.kt

```diff
@@ -95,17 +116,47 @@ data class DetailContext(
     val isTruncatedOnRight: Boolean,
     val rightOverlap: String = "",
     val completionType: CodewhispererCompletionType,
+    var hasSeen: Boolean = false,
+    var isAccepted: Boolean = false
 )
 
 data class SessionContext(
-    val typeahead: String = "",
-    val typeaheadOriginal: String = "",
-    val selectedIndex: Int = 0,
+    val project: Project,
+    val editor: Editor,
+    var popup: JBPopup? = null,
+    var selectedIndex: Int = -1,
     val seen: MutableSet<Int> = mutableSetOf(),
-    val isFirstTimeShowingPopup: Boolean = true,
+    var isFirstTimeShowingPopup: Boolean = true,
     var toBeRemovedHighlighter: RangeHighlighter? = null,
-    var insertEndOffset: Int = -1
-)
+    var insertEndOffset: Int = -1,
+    var popupDisplayOffset: Int = -1,
+    val latencyContext: LatencyContext,
+    var hasAccepted: Boolean = false
+) : Disposable {
+    private var isDisposed = false
+
+    @RequiresEdt
+    override fun dispose() {
+        CodeWhispererTelemetryService.getInstance().sendUserDecisionEventForAll(
+            this,
+            hasAccepted,
+            CodeWhispererInvocationStatus.getInstance().popupStartTimestamp?.let { Duration.between(it, Instant.now()) }
+        )
+        CodeWhispererInvocationStatus.getInstance().setDisplaySessionActive(false)
+
+        if (hasAccepted) {
+            popup?.closeOk(null)
+        } else {
+            popup?.cancel()
+        }
+        popup?.let { Disposer.dispose(it) }
+        popup = null
```

Comment:
popup manager and sessioncontext are so tied together that it feels like it should just be handled together

Reply:
popup now becomes a state of a session (1 session 1 popup) so I put it as a field of `SessionContext`, and then I let disposing SessionContext handles everything so I don't have to call 
```
// CWService.getInstance().disposeDisplaySession()
// popup.cancel()/close()
```
separately.